### ISSUE TITLE
V8: Magic strings doesn't display value

### ISSUE THREAD

**Issue by Frost117** [umbraco/Umbraco.Forms.Issues#176]
If you want to display the [#CurrentPage] for when the forms are being used, then you won't be able to do it in version 8.

In the documentation we have the following:
https://our.umbraco.com/documentation/Add-ons/UmbracoForms/Developer/Magic-Strings/#umbraco-page-field-item

But if you want to use this in a forms you will be left with nothing rendered in the view.

This is what it looks like in version 8:
![image](https://user-images.githubusercontent.com/32164310/60077784-05e82a80-972b-11e9-9158-b574374ddf5e.png)

Which should render out the name "Root Node".
![image](https://user-images.githubusercontent.com/32164310/60077755-f79a0e80-972a-11e9-915f-826cc03afae3.png)

And as you can see it fails to do so.

You are able to do it version 7.

![image](https://user-images.githubusercontent.com/32164310/60077875-30d27e80-972b-11e9-80db-0b4b18f1a9cb.png)
![image](https://user-images.githubusercontent.com/32164310/60077925-447de500-972b-11e9-876b-8ce0aa7c3227.png)

The exact same setup, with 2 different results.

Reproduction
------------

### Steps to reproduce

1. make form in v8. 
2. In the label put in [#pageName] like so: 
![image](https://user-images.githubusercontent.com/32164310/60077784-05e82a80-972b-11e9-9158-b574374ddf5e.png)

3. Make a doctype, with a template, that has a form picker.
4. Make a content node using the doctype that you have just made.
If the form doesn't get displayed in the front-end you will need to update the template so it renders out the picked form.

5. If you visit the front-end of the site you should see that the [#pageName] does not render out what you called the content node.

### Specifics
The problem occurs on the root node.

Umbraco 8.0.1
Forms 8.1.1
Deploy 3.0.2

Problem exist in Firefox, Chrome and Edge

### Expected result

That the form label displays the current node it is being viewed on.

### Actual result

Nothing gets displayed.

---
_This item has been added to our backlog [AB#2334](https://umbraco.visualstudio.com/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/2334)_

**Comment by facingdigital** [comment]
I have the same problem to use the magic string in the workflow step...
I have a form field with an alias and a caption "topic" 
![Screenshot_2019-06-28 (Local) Forms - localhost](https://user-images.githubusercontent.com/41913343/60350820-b3c23600-99c4-11e9-9d29-db06d6dfb522.png)
![Screenshot_2019-06-28 (Local) Forms - localhost(1)](https://user-images.githubusercontent.com/41913343/60350827-b6249000-99c4-11e9-99dc-5e46694d351c.png)
![Screenshot_2019-06-28 (Local) Forms - localhost(2)](https://user-images.githubusercontent.com/41913343/60350831-b7ee5380-99c4-11e9-8a73-cd6327b632aa.png)

I put the value {topic} in the workflow field "receiver email" but this value is empty....

![Screenshot_2019-06-28 (Local) Settings - localhost](https://user-images.githubusercontent.com/41913343/60350958-04399380-99c5-11e9-8af6-8822ea5eaf9c.png)

Please, it's very urgent to fix this big issue!

Thank you for your comprehension.

User from Umbraco Cloud

**Comment by shaishavkarnani** [comment]
Did this issue get fixed? I can see same issue in Umbraco version 8.1.1 and Umbraco Form version 8.1.1

**Comment by rgdagostino** [comment]
Same issue here on Umbraco 8.1.2 and Forms 8.1.1 
![1](https://user-images.githubusercontent.com/8196233/63148908-efcd5b00-c045-11e9-8ce0-6114feecfa2e.PNG)
![2](https://user-images.githubusercontent.com/8196233/63148914-f2c84b80-c045-11e9-8e6f-49ea6c23d0fc.PNG)

**Comment by seekingcode18** [comment]
I'm having the same issue.
Umbraco 8.1.2
Forms 8.1.1
Deploy 3.1.1

**Comment by callumbwhyte** [comment]
Also having this issue!

Umbraco 8.1.2
Forms 8.1.1

Given this seems to be affecting a few people I hope we can get a resolution soon 🙏

**Comment by nul800sebastiaan** [comment]
Looks like in v7 we were relying on some things we insterted into `HttpContext.Items` from withing Umbraco, items like `pageName` and `pageID` used to be available there, this is no longer the case in v8 making these two items fail.

Is there other magic strings than those two that are currently failing? It looks like everything else is wired up correctly. We can probably hack these back in to Forms as special cases. We'll have a look.

**Comment by callumbwhyte** [comment]
That was exactly as my investigation found too - as the "pageElements" hashtable is no longer in HTTP Context none of the data needed is there.

Previously we could do `[#myPropertyAlias]` to get any value from the current content item, or `[$myPropertyAlias]` to get a value recursively. This logic relied on the `pageID` from that "pageElements" collection, which no longer works (as above)...

The big question here is how do we deal with variant properties? This applies to both custom properties, and the page name... My (selfish) preference would be to get the value for the "default" language and ignore other variants for now as I think this handling this properly is a much bigger discussion!

**Comment by clausjensen** [comment]
@callumbwhyte I'd say it would make sense to do what we also do if you fetch a property without specifying a variant in a normal template. We would pick the variant data that fits with the current context being rendered.

The reason being that if you use #pageName (without specifying a culture) in your form rendering - you'd get 'Page Name' displayed. You render the same field out in your template - you get 'Page Name'.
If you then switch to a context of Danish .. you would have 'Side Navn' rendered in the template due to the culture being changed - but just using default language for Forms would then cause 'Page Name' to be rendered in the form rendering.

I don't think we (at the moment) should support fetching fields from different cultures for these magic strings. Not sure if anyone would ever need that so until then let's not try to do that.

**Comment by callumbwhyte** [comment]
Hey FRIENDLY® Forms team 😊

Just wondering if there was any progress on getting this fixed? Or anything we can do as the community to help you out...?

Understand there's quite a few bugs to triage / fix with Forms, and there's other priorities, but it's disappointing to see "minor" fixes for more recent issues than this one being fixed and silence here when this is affecting quite a few people... ❤

**Comment by Robiwahn23** [comment]
Yeah, any reaction from the forms team would be nice!

**Comment by NikRimington** [comment]
Hey all,

I've got the same issue with place holders relating to the same node. However you can hack it around a little bit and use this inside what ever calls your form render, be it a macro or a partial to put the basics in there:

```
var hashTable = new System.Collections.Hashtable{ {"pageName", Umbraco.AssignedContentItem.Name}, {"pageID", Umbraco.AssignedContentItem.Id}, {"pageKey", Umbraco.AssignedContentItem.Key} };
    if(Context.Items.Contains("pageElements") == false)
    {
        Context.Items.Add("pageElements", hashTable);
    }
```

Although I've not done it here, if you also add a property with the key "splitpath" that is a string[] of the path value split on `,` then you'll also be able to use the $ place holders for properties on the document type. I think that it will the current culture if using variants, although I've not tested this to be honest.

**Comment by NikRimington** [comment]
Since my comment yesterday I've tested adding in splitpath and that seems to work as well, giving access to the content item.

**Comment by callumbwhyte** [comment]
I ended up doing the same as @NikRimington, but certainly not an ideal solution as editors can place forms in many places on the site... A "proper" solution would be great 😊

Wondering if there's a nice-ish way of running this code on every request? Was thinking I could do this by overriding something in `UmbracoApplication` but couldn't see anywhere suitable...

**Comment by bergmania** [comment]
Hi all..

The issue is because umbraco 8, does not add the `pageElements` to the `HttpContext.Items` like Umbraco 7 did.

I consider a fix in forms that add these `pageElements` to `HttpContext.Items` when the form model is build. This will fix it, but if there are two forms on the same page, some redundant work is done.

**Comment by warrenbuckley** [comment]
Fixed in this PR that's merged in now
https://github.com/umbraco/Forms/pull/305

**Comment by vmashuryan** [comment]
Hi,

I'm on Umbraco Cloud and on Umbraco 8.1.5, Forms 8.2.0. Magic strings are not working at all for me, it always populates empty values for both form fields and page level fields. Though it populates some predefined variables such as @Http_Referer, or form_name, all the form fields and page level fields are coming as empty. I'm trying to get values from both from design as default values and from Razor email template using ParsePlaceHolders. Support suggested to re-open this case.

**Comment by bergmania** [comment]
Hi @vmashuryan 

I just tried to make a form with a field with alias `email` and used the workflow of type `Send email with template (Razor)`

I set the Email field of the workflow to  `{email}`, and it seems to work. Also `{form_name}` is replaced in my example.

Are you able to give a concrete step-by-step guide for reproduction?

**Comment by vmashuryan** [comment]
Hi @bergmania ,

Thanks for the update. Steps are pretty much simple. I have a job posting form and in that page there is a contactEmail ( alias ) field, which is unique for each location. So different forms receiver emails should be different. I'm using [#contactEmail], [$contactEmail], {contactEmail}, {contactemail}, etc.. There are other fields as well, however they are not being populated for both form design ( as a default value ) and email template ( custom Razor). Strangest thing is that even form fields are not being populated for example if I use {firstName} in template. Furthermore it is not being populated even for Success Message. 
Is it possible that issue is because I've created those forms in previous Forms ( where issue was known and not fixed yet ) , so I have to re-create them?

**Comment by vmashuryan** [comment]
I sent your requested file, as well as some screenshots. Let me know if you need anything else:

Thanks

**Comment by bergmania** [comment]
@vmashuryan's issue was not related to this issue, so I close the issue again.

**Comment by joepvtl** [comment]
Hi, when you insert a form as macro into an richtexteditor and it tries to render the macro in the backoffice. It gives a nullref exception since you don't have the publishedrequest available in the backoffice (tried on Umbraco 8.2, 8.3 and 8.4 all with the latest version of forms).
